src/ConnectionPool.jl

```diff
@@ -0,0 +1,575 @@
+Remove closed connections from `pool`.
+"""
+function purge()
+    while (i = findfirst(x->!isopen(x.io), pool)) > 0
+        c = pool[i]
+        deleteat!(pool, i)                    ;@debug 1 "🗑  Deleted:        $c"
+    end
+end
+
+
+"""
+    getconnection(type, host, port) -> Connection
+
+Find a reusable `Connection` in the `pool`,
+or create a new `Connection` if required.
+"""
+
+function getconnection(::Type{Transaction{T}},
+                       host::AbstractString,
+                       port::AbstractString;
+                       connection_limit::Int=default_connection_limit,
+                       pipeline_limit::Int=default_pipeline_limit,
+                       reuse_limit::Int=nolimit,
+                       kw...)::Transaction{T} where T <: IO
+
+    while true
+
+        lock(poollock)
+        try
+
+            # Close connections that have reached the reuse limit...
+            if reuse_limit != nolimit
+                for c in findoverused(T, host, port, reuse_limit)
+                    close(c)
+                end
+            end
+
+            # Remove closed connections from `pool`...
+            purge()
+
+            # Try to find a connection with no active readers or writers...
+            writable = findwritable(T, host, port, pipeline_limit, reuse_limit)
+            idle = filter(c->!c.readbusy, writable)
+            if !isempty(idle)
+                c = rand(idle)                 ;@debug 2 "♻️  Idle:           $c"
+                return client_transaction(c)
+            end
+
+            # If there are not too many connections to this host:port,
+            # create a new connection...
+            busy = findall(T, host, port, pipeline_limit)
+            if length(busy) < connection_limit
+                io = getconnection(T, host, port; kw...)
+                c = Connection(host, port, pipeline_limit, io)
+                push!(pool, c)                ;@debug 1 "🔗  New:            $c"
+                return client_transaction(c)
+            end
+
+            # Share a connection that has active readers...
+            if !isempty(writable)
+                c = rand(writable)             ;@debug 2 "⇆  Shared:         $c"
+                return client_transaction(c)
+            end
+
+        finally
+            unlock(poollock)
+        end
+
+        # Wait for `closewrite` or `close` to signal that a connection is ready.
+        wait(poolcondition)
+    end
+end
+
+
+function getconnection(::Type{TCPSocket},
+                       host::AbstractString,
+                       port::AbstractString;
+                       kw...)::TCPSocket
+
+    p::UInt = isempty(port) ? UInt(80) : parse(UInt, port)
+    @debug 2 "TCP connect: $host:$p..."
+    connect(getaddrinfo(host), p)
+end
+
+
+function getconnection(::Type{SSLContext},
+                       host::AbstractString,
+                       port::AbstractString;
+                       require_ssl_verification::Bool=false,
```

Comment:
seems like the default here should be `true`

Reply:
cb9ad328c7348d811147c1bd511a5b5be3b0fb04